### ISSUE TITLE
Preserve breakpoints across debuggee restart

### ISSUE THREAD

**Issue by 10110111** [eteran/edb-debugger#358]
Pressing Ctrl+F2 (or choosing Debug->Restart) currently is (mostly) equivalent to killing the process and launching it again. In particular, the breakpoints are lost on such restart. It'd be useful if they could be preserved.

**Comment by eteran** [comment]
Right. Now that we have session files this is doable, but has a little bit of trickiness to it:

1. ASLR, we need to make sure to store breakpoints as offsets from a region's start instead of absolute addresses. We probably should store a hash of each region that has a BP so we can detect if it changed and therefore should not restore BPs.

2. This is the trickier part. Not all libraries are loaded on start, and with the ptrace API, there is no "event" for library was just loaded like there is on Windows... But there is a solution.

By the time we get to "main", there is a "debug pointer" that is set by ld (see `Debugger::next_debug_event` for some code which attempts to identify it).

One of the things that this pointer refers to is a place in memory which will be jumped to via a `call` instruction on each library load event. So we need to place an internal BP there and treat hitting that as a library load event. It's not pretty, but it'll work :-).

**Comment by eteran** [comment]
So once we have the library load hooked, my current thoughts for how to store breakpoints would look like this:

    {
        [
            "module" : "/lib64/ld-2.21.so",
            "checksum" : "<MD5>",
            "breakpoint_offsets" : [ "1234", "5678", "9abc" ... ]
        ]
    }

Addresses in hex as strings so we don't lose precision converting between floats and values. This would support multiple regions and their offsets and we can detect if the module changed (letting us warn during restoration).

I think this'll cover our needs, but first thing's first, gotta finish implementing load library hooks on linux.

**Comment by 10110111** [comment]
What do you mean by floats? Addresses are integers.

**Comment by eteran** [comment]
Exactly. But JSON numbers according to the JSON standard are floating point. So to prevent any unexpected loss of precision, we should store all addresses as strings and convert back to integer values when we read it back.

**Comment by ojura** [comment]
+1 for adding this.

**Comment by Anutrix** [comment]
Any updates on this?

**Comment by redthing1** [comment]
Any update on this?

**Comment by eteran** [comment]
It's been a while for sure, I would love to get this done, but life has been particularly busy. We'll get there eventually though :-).

**Comment by eteran** [comment]
@CherryDT (and others), I certainly want to get this in there. I've even been considering an improved session API which would make use of sqlite DB instead of just storing a JSON blob.

I hate to say it, and I know it's been forever, but "life has been busy" is still the case (mostly really good stuff like Marriage, friends, vacations, work, etc...) . I hope to get this feature (and others!) in there sometime soon.

I appreciate your offer to support financially, and look, I'm never going to turn down a donation... and I certainly would feel motivated to set aside SOME time to work on edb in return... but to be completely transparent, one of the (numerous) things that keeps me very busy lately is work, so when it comes to financial incentives, that is typically going to win.

I'll do my best to work on things more.

**Comment by eteran** [comment]
The needle has recently moved a bit on this!

This past week I revamped how we detect library load/unload events and it is MUCH more reliable than it was, This should lay a strong foundation for save/restore breakpoints/comments/etc in libraries which was the big hurdle in my mind. I'll keep you guys posted.

**Comment by eteran** [comment]
Implemented! (and ASLR aware)
